pkg/registry/populator.go

```diff
@@ -26,124 +27,91 @@ type DirectoryPopulator struct {
-	}
-
-	// Look for the metadata and manifests sub-directories to find the annotations.yaml
-	// file that will inform how the manifests of the bundle should be loaded into the database.
-	// If dependencies.yaml which contains operator dependencies in metadata directory
-	// exists, parse and load it into the DB
-	annotationsFile := &AnnotationsFile{}
-	dependenciesFile := &DependenciesFile{}
-	for _, f := range files {
-		err = decodeFile(filepath.Join(metadata, f.Name()), annotationsFile)
-		if err != nil || *annotationsFile == (AnnotationsFile{}) {
-			log.Info("found annotations file searching for csv")
+	var errs []error
+	imagesToAdd := make([]*ImageInput, 0)
+	for to, from := range i.imageDirMap {
+		imageInput, err := NewImageInput(to, from)
+		if err != nil {
+			errs = append(errs, err)
 			continue
 		}
 
-		err = parseDependenciesFile(filepath.Join(metadata, f.Name()), dependenciesFile)
-		if err != nil || len(dependenciesFile.Dependencies) < 1 {
-			continue
-		} else {
-			log.Info("found dependencies file searching for csv")
-		}
+		imagesToAdd = append(imagesToAdd, imageInput)
 	}
 
-	if *annotationsFile == (AnnotationsFile{}) {
-		return fmt.Errorf("Could not find annotations.yaml file")
+	if len(errs) > 0 {
+		return utilerrors.NewAggregate(errs)
 	}
 
-	err = i.loadManifests(manifests, annotationsFile, dependenciesFile, mode)
+	err := i.loadManifests(imagesToAdd, mode)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (i *DirectoryPopulator) loadManifests(manifests string, annotationsFile *AnnotationsFile, dependenciesFile *DependenciesFile, mode Mode) error {
-	log := logrus.WithFields(logrus.Fields{"dir": i.from, "file": manifests, "load": "bundle"})
-
-	csv, err := i.findCSV(manifests)
-	if err != nil {
-		return err
-	}
-
-	if csv.Object == nil {
-		return fmt.Errorf("csv is empty: %s", err)
-	}
-
-	log.Info("found csv, loading bundle")
-
-	csvName := csv.GetName()
-
-	bundle, err := loadBundle(csvName, manifests)
-	if err != nil {
-		return fmt.Errorf("error loading objs in directory: %s", err)
-	}
-
-	if bundle == nil || bundle.Size() == 0 {
-		return fmt.Errorf("no bundle objects found")
-	}
-
-	// set the bundleimage on the bundle
-	bundle.BundleImage = i.to.String()
-	// set the dependencies on the bundle
-	bundle.Dependencies = dependenciesFile.GetDependencies()
-
-	bundle.Name = csvName
-	bundle.Package = annotationsFile.Annotations.PackageName
-	bundle.Channels = strings.Split(annotationsFile.Annotations.Channels, ",")
-
-	if err := bundle.AllProvidedAPIsInBundle(); err != nil {
-		return fmt.Errorf("error checking provided apis in bundle %s: %s", bundle.Name, err)
-	}
-
+func (i *DirectoryPopulator) loadManifests(imagesToAdd []*ImageInput, mode Mode) error {
 	switch mode {
 	case ReplacesMode:
-		err = i.loadManifestsReplaces(bundle, annotationsFile)
-		if err != nil {
-			return err
+		// TODO: This is relatively inefficient. Ideally, we should be able to use a replaces
```

Comment:
This comment is also my main comment - we have a lot of the lower level stuff in place to do this really efficiently now. 

But I don't think that needs to block

Reply:
yeah, looking at it it was just going to take a lot more time and effort to reimplement some of the sql layer to do that, so I opted for this initial implementation to get it out there and kicked the can down the road somewhat.

as for a test, not explicitly called out, but the root db populate now generates the database with out of order input: https://github.com/operator-framework/operator-registry/pull/285/files#diff-bac4eb64785fb97243acdc492eedd162R70